content/blog/2022-05-19-fast-brili.md

```diff
@@ -0,0 +1,141 @@
++++
+title = "Bril bytecode interpreter"
+[extra]
+bio = """
+	Charles Sherk is an junior studying Computer Science & Mathematics at
+	Cornell university, interested in compilers and programming language
+	implementation.
+	Susan Garry is a senior studying Computer Science at Cornell University,
+	interested in compilers and programming language design.
+
+"""
+[[extra.authors]]
+name = "Charles Sherk"
+link = "https://github.com/charles-rs/"
+[[extra.authors]]
+name = "Susan Garry"
+link = "https://github.com/susan-garry"
++++
+
+## Motivation
+Currently we have access to `brili`, which is written with an emphasis on
+readable code and extensiblity, and `brili-rs`, which is supposed to be faster,
+but `brili-rs` isn't actually that fast, and seems to be rather bloated for what
+it is. There were some points in this course where time was spent on "waiting
+for the Bril interpreter" that could have been avoided. So, we propose and
+implement an interpreter for Bril whose sole focus is speed.
+
+## Design overview
+
+### Language choice
+We chose C for an implementation language as it is known for its speed as well
+as simplicity, which means we won't inadvertently write code that is more
+expensive than necessary.
+
+### Big Red Bytecode
+In order to make our interpreter as efficient as possible, we have designed a
+very compressed representation of bril instructions: (almost) all of them are
+represented in 64 bits, with some taking multiple 64 bit words as needed. See
+the [documentation](https://github.coecis.cornell.edu/cs897/fast-bril/blob/master/doc/brb.pdf).
+
+In order to appease the C typechecker and not rely on undefined behavior, we use
+a union type to internally represent instructions, but most have the first 16
+bits as the opcode, so we can pull this out and then determine how to decode the
+rest of the instruction. In principle, we could also use `uint64_t` or
+`uint16_t[4]`, but using unions is the most user friendly, and lets the C
+compiler do more work to access the fields of structs, in case there are any
+cool tricks for that behind the scenes.
+
+### Features
+We support a superset of the behavior provided by `brili`, so options like `-p`
+work exactly the same. We also support the following:
+ - `-b` will read in bytecode instead of the Bril json
+ - `-bo <file>` will output the bytecode to `<file>`
+ - `-pr` will print the program to standard out (probably more useful with the
+         `-b` option)
+ - `-ni` will NOT run the interpreter.
+ - `-e <file>` will emit assembly to `<file>`.
```

Comment:
I know this is too low-level for a blog post about the virtues of bytecode interpretation, but I am still confused after reading this list about the combination of flags necessary to accomplish some stuff, like this:

- Translate JSON to bytecode. Probably `fastbrili -bo < foo.json`? Maybe I have to add `-ni`? Maybe I need an output file?
- Translate JSON to JSON. Probably `fastbrili -pr < foo.json`? Also maybe needs `-ni`? Or maybe it *also* needs `pr`? Not sure.
- Interpret bytecode. Maybe `fastbrili -b < foo.bc`??

In other words, I think these options could be made a bit more orthogonal…

"Emit assembly" is also a perplexing to see here. So far we've been talking about interpretation. It also emits assembly? If so, you should say this sooner, or maybe later in the compiler section, or something!

Reply:
`fastbrili -bo < foo.json` will print bytecode to the file `my-output`, and run the interpreter
`fastbrili `-bo` foo.bc < foo.json` will print bytecode to the file `foo.bc`, and run the interpreter

`fastbrili -pr < foo.json` will print foo's program to stdout, and run the interpreter. it's the text representation not the json right now... this was initially a debugging feature, but i really should change it to json

spot on for interpreting bytecode.

perhaps options that cause "something else" to happen should disable the interpreter by default, removing the `-ni` flag altogether? probably not that often someone wants to interpret and compile all in one go